src/Microsoft.ML.AutoML/API/ExperimentBase.cs

```diff
@@ -67,11 +67,23 @@ public abstract class ExperimentBase<TMetrics, TExperimentSettings>
         public ExperimentResult<TMetrics> Execute(IDataView trainData, string labelColumnName = DefaultColumnNames.Label,
             string samplingKeyColumn = null, IEstimator<ITransformer> preFeaturizer = null, IProgress<RunDetail<TMetrics>> progressHandler = null)
         {
-            var columnInformation = new ColumnInformation()
+            ColumnInformation columnInformation;
+            if (_task == TaskKind.Ranking)
             {
-                LabelColumnName = labelColumnName,
-                SamplingKeyColumnName = samplingKeyColumn
-            };
+                columnInformation = new ColumnInformation()
+                {
+                    LabelColumnName = labelColumnName,
+                    GroupIdColumnName = samplingKeyColumn ?? DefaultColumnNames.GroupId
+                };
```

Comment:
**Suggestion:** As per the feedback we got from @justinormont  today, I think it would be better to set _both_ `GroupIdColumnName` and `SamplingKeyColumnName` in here. Something like:

```C#
columnInformation = new ColumnInformation()
{
    LabelColumnName = labelColumnName,
    SamplingKeyColumnName = samplingKeyColumn ?? DefaultColumnNames.GroupId,
    GroupIdColumnName = samplingKeyColumn ?? DefaultColumnNames.GroupId // For ranking, we want to enforce having the same column as samplingKeyColum and GroupIdColumn
}
```

With your current implementation it won't make any difference to do this, but I do think this might be clearer for future AutoML.NET developers.

A similar change would need to take place in the other overload that receives a `samplingKeyColumnName` but no `columnInformation`.

Reply:
I would lean towards deferring this to the next update. I will take a quick look, but having a column with two column info seems to be causing issues.

Reply:
It's just a two line change (adding the line in here, and in the other overload), and it's just to have it clear in the columnInformation object that we'll be using the `samplingKeyColumn` provided by the user _both_ as `SamplingKeyColumnName` and `GroupIdColumnName` (which is actually what we're doing). So I think it's clearer this way. But whatever you decide is fine :wink:

---
In reply to: [452987600](https://github.com/dotnet/machinelearning/pull/5263#discussion_r452987600) [](ancestors = 452987600)

Reply:
Just to be clear, mapping a groupId column to both `SamplingKeyColumnName` and `GroupIdColumnName` doesn't work with the current implementation. The current implementation uses `GroupIdColumnName` as the `SamplingKeyColumnName`, so if the user provides a `SamplingKeyColumnName,` we throw an error (unless they are both the same).

Reply:
Yeah, I know the current implementation throws if they're not the same. That's way I suggested using the `samplingKeyColumn` to set _both_ `SamplingKeyColumnName` and `GroupIdColumnName`.

In general if the user provides a `ColumnInformation` object containing `SamplingKeyColumnName` and `GroupIdColumnName`, then we should accept that if both are the same (and in the current implementation this is doable). So I'm just not sure what's the problem in here.

---
In reply to: [453055824](https://github.com/dotnet/machinelearning/pull/5263#discussion_r453055824) [](ancestors = 453055824)